### ISSUE TITLE
output_attentions gating and model_kwargs_method update strategy in AVISC sampling for data generation

### ISSUE THREAD

**Issue by Anyna918** [LijunZhang01/Octopus#3]
Thank you for your great work and i'm really love it!!! But I have a question. I noticed an internal contradiction in the Octopus sampling code that suggests the AVISC intervention is only effective at the first generation step, despite the paper claiming it operates at every token generation step.

The contradiction:
On one hand, output_attentions is gated to only be True at the first step (when past_key_values is None), which correctly reflects that the blind token mask is only computed once:
`output_attentions = use_avisc and not (
    model_kwargs.get("use_cache") and model_kwargs.get("past_key_values") is not None
)`

On the other hand, model_kwargs_method is updated using the main branch outputs rather than outputs_method:
`model_kwargs_method = self._update_model_kwargs_for_generation(
    outputs, model_kwargs_method, ...  # should this be outputs_method?
)`

The consequence:
From step 2 onward, model_kwargs_method carries the exact same KV cache as the main branch. Since past_key_values is no longer None, mask_idx is set to None, and the auxiliary forward becomes identical to the main forward. 

The question:
Was the intended design to follow the original AVISC paper — maintaining a separate masked KV cache in the auxiliary branch across all steps (which would require updating model_kwargs_method with outputs_method)? Or is the first-step-only approximation intentional for efficiency reasons?

**Comment by LijunZhang01** [comment]
Thank you for your attention and for raising this question. I may not have fully understood your concern, but if I understood correctly, you are asking whether AVISC is applied only once. That would indeed be incorrect. In our implementation, AVISC can be invoked at every token generation step.

I briefly checked the code you mentioned. In our training/inference pipeline, the relevant implementation is in `avisc_sample.py`. At line 357, `model_kwargs_method` is indeed updated with `outputs_method`:

```python
model_kwargs_method = self._update_model_kwargs_for_generation(
                outputs_method, model_kwargs_method, is_encoder_decoder=self.config.is_encoder_decoder
            )
```

So it is **not** updated with `outputs`.

What may have caused the confusion is the corresponding line in `avisc_sample_gendata.py`. That file was added later in response to an issue, and it contains the final code we uploaded for the **discriminative task**. For the discriminative setting, considering only the first token is sufficient, so using that version of the code is reasonable.

It has been quite a while, so I am a bit rusty with some parts of the code myself. I am not entirely sure whether I have fully understood and answered your question, but I hope this helps. If anything is still unclear, please feel free to email us.

Thank you again for your question.

**Comment by Anyna918** [comment]
Thanks for the explanation! I'd like to ask about the part regarding the generation of training data for DPO. I was wondering if avisc_sample_gendata.py is the script used to generate this data? Or is it only used for generating training data for discriminative tasks, or perhaps just for inference?

**Comment by LijunZhang01** [comment]
It is used to generate data for discriminative tasks, and with slight modifications, it can also be used to generate tasks.

**Comment by Anyna918** [comment]
Okay！Thanks for your detailed explanation！Have a nice weekend！
